### PR TITLE
Backend/feat: #916/ Driver Rating Customer

### PR DESCRIPTION
**Pull request by vsaimanohar**
Driver Rating Customer

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Driver Rating Customer

### Additional Changes

- [X] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [X] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [X] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by umangj-28 — APPROVED** [review]

File: Backend/dev/migrations/rider-app/1116-rating.sql [hunk 1]
```diff
@@ -0,0 +1,12 @@
+CREATE TABLE atlas_app.rating (
+id character(36) NOT NULL,
+ride_id character varying(36) NOT NULL,
+rating_value bigint,
+feedback_details character varying(255),
+rider_id character varying(36) NOT NULL,
```
Comment (JanhaviSoni) [inline Backend/dev/migrations/rider-app/1116-rating.sql:6]: ride_id and rider_id should have character(36)
Should rating_value be bigint ? What is its range

File: Backend/dev/migrations/rider-app/1116-rating.sql [hunk 2]
```diff
@@ -0,0 +1,12 @@
+CREATE TABLE atlas_app.rating (
+id character(36) NOT NULL,
+ride_id character varying(36) NOT NULL,
+rating_value bigint,
```
Comment (JanhaviSoni) [inline Backend/dev/migrations/rider-app/1116-rating.sql:4]: rating value is NULL so it should be of maybe type

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Tabular/Rating.hs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import Storage.Tabular.Ride (RideTId)
+
+mkPersist
+  defaultSqlSettings
+  [defaultQQ|
+    RatingT sql=rating
+      id Text
+      rideId RideTId
+      riderId PersonTId
+      ratingValue Int
```
Comment (JanhaviSoni) [inline Backend/app/rider-platform/rider-app/Main/src/Storage/Tabular/Rating.hs:37]: maybe type

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Rating.hs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+type API =
+  "rating"
+    :> TokenAuth
+    :> ReqBody '[JSON] RatingReq
+    :> Post '[JSON] APISuccess
+
+handler :: FlowServer API
+handler =
+  getDriverRating
```
Comment (JanhaviSoni) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Rating.hs:40]: this name can be something which specifies that it is customer's rating which driver is giving. It can be misleading as rating of the driver. Can be something like getCustomerRatingByDriver

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Rating.hs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import Domain.Types.Person (Person)
+import Domain.Types.Ride (Ride)
+import EulerHS.Prelude hiding (id)
+import Kernel.Types.Id
+
+data Rating = Rating
+  { id :: Id Rating,
+    rideId :: Id Ride,
+    riderId :: Id Person,
+    ratingValue :: Int,
```
Comment (JanhaviSoni) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Rating.hs:27]: ratingValue --> maybe type

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Rating.hs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import Environment
+import EulerHS.Prelude
+import Kernel.Types.APISuccess
+import Kernel.Types.Id (Id (..))
+import Kernel.Utils.Common
+import Servant
+import Tools.Auth
+
+type API =
+  "rating"
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Rating.hs:33]: `feedback/rateRide`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Rating.hs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    :> TokenAuth
+    :> ReqBody '[JSON] RatingReq
+    :> Post '[JSON] APISuccess
+
+handler :: FlowServer API
+handler =
+  getCustomerRatingByDriver
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Rating.hs:40]: `feedback`

File: Backend/lib/beckn-spec/src/Beckn/Types/Core/Taxi/OnUpdate/OnUpdateEvent/DriverRatingEvent.hs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+import Data.OpenApi hiding (Example, example, title, value)
+import EulerHS.Prelude hiding (id, (.=))
+import GHC.Exts (fromList)
+import Kernel.Utils.Schema
+
+data DriverRatingEvent = DriverRatingEvent
+  { id :: Text,
+    update_target :: Text,
+    fulfillment :: FulfillmentInfo,
+    rating_value :: Int,
```
Comment (hkmangla) [inline Backend/lib/beckn-spec/src/Beckn/Types/Core/Taxi/OnUpdate/OnUpdateEvent/DriverRatingEvent.hs:34]: Add rating_value and `feedback_details` inside the tags instead of adding here.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Rating.hs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  { rideId :: Id Ride,
+    ratingValue :: Int,
+    feedbackDetails :: Maybe Text
+  }
+  deriving (Generic, Show, FromJSON, ToJSON, ToSchema)
+
+getCustomerRatingByDriver :: (EncFlow m r, CacheFlow m r, EsqDBFlow m r, EsqDBReplicaFlow m r, CoreMetrics m, HasFlowEnv m r '["nwAddress" ::: BaseUrl], HasHttpClientOptions r c, HasShortDurationRetryCfg r c) => RatingReq -> m APISuccess
+getCustomerRatingByDriver req = do
+  ride <- runInReplica $ QRide.findById req.rideId >>= fromMaybeM (RideDoesNotExist req.rideId.getId)
+  unless (isValidRideStatus (ride.status)) $ throwError $ RideInvalidStatus "The ride has not completed yet."
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Rating.hs:40]: why extra bracket on `ride.status`?

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Rating.hs (hunk 5 above)
Comment (hkmangla) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Rating.hs:27]: `rideId` can be the primaryKey instead of creating one more id `id`

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Queries/Rating.hs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  Esq.update $ \tbl -> do
+    set
+      tbl
+      [ RatingRatingValue =. val newRatingValue,
+        RatingFeedbackDetails =. val newFeedbackDetails,
+        RatingUpdatedAt =. val now
+      ]
+    where_ $
+      tbl ^. RatingTId ==. val (toKey ratingId)
+        &&. tbl ^. RatingRiderId ==. val (toKey riderId)
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/src/Storage/Queries/Rating.hs:41]: No need to do check for riderId.

File: Backend/dev/migrations/rider-app/1116-rating.sql (hunk 2 above)
Comment (hkmangla) [reply Backend/dev/migrations/rider-app/1116-rating.sql:4]: using int instead of bigint

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Rating.hs (hunk 5 above)
Comment (vsaimanohar) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Rating.hs:27]: can we do the same change for Customer Rating Table too @hkmangla?

Comment (vsaimanohar) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Types/Rating.hs:27]: @hkmangla can we make ratingId itself as unique identifier for this table instead of rideId to align ourself with normalization rules

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FeedBack.hs [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    feedbackDetails :: Maybe Text
+  }
+  deriving (Generic, Show, FromJSON, ToJSON, ToSchema)
+
+feedback :: (EncFlow m r, CacheFlow m r, EsqDBFlow m r, EsqDBReplicaFlow m r, CoreMetrics m, HasFlowEnv m r '["nwAddress" ::: BaseUrl], HasHttpClientOptions r c, HasShortDurationRetryCfg r c) => FeedbackReq -> m APISuccess
+feedback request = do
+  unless (request.ratingValue `elem` [1 .. 5]) $ throwError InvalidRatingValue
+  let rideId = request.rideId
+  ride <- QRide.findById rideId >>= fromMaybeM (RideDoesNotExist rideId.getId)
+  unless (ride.status == COMPLETED) $ throwError (RideInvalidStatus "Feedback available only for completed rides.")
```
Comment (DonDokidon) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FeedBack.hs:43]: I think it should be possible for CANCELLED too?

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Tabular/Rating.hs (hunk 3 above)
Comment (DonDokidon) [reply Backend/app/rider-platform/rider-app/Main/src/Storage/Tabular/Rating.hs:35]: > maybe type

for what? Why someone might want to rate ride as null?

File: Backend/dev/migrations/rider-app/1127-rating.sql [hunk 12]
```diff
@@ -0,0 +1,12 @@
+CREATE TABLE atlas_app.rating (
+id character(36) NOT NULL,
+ride_id character varying(36) NOT NULL REFERENCES atlas_app.ride (id),
+rating_value int,
```
Comment (DonDokidon) [inline Backend/dev/migrations/rider-app/1127-rating.sql:4]: NOT NULL if it is not null in the actual code

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FeedBack.hs (hunk 11 above)
Comment (vsaimanohar) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/FeedBack.hs:43]: for now we are only doing it for COMPLETED Status . cc :: @gouri-muley

File: Backend/dev/migrations/rider-app/1127-rating.sql (hunk 12 above)
Comment (vsaimanohar) [reply Backend/dev/migrations/rider-app/1127-rating.sql]: feedback_details column is maybe so we can give null

File: Backend/app/rider-platform/rider-app/Main/src/Environment.hs [hunk 13]
```diff
@@ -101,6 +101,7 @@ data AppCfg = AppCfg
     rideCfg :: RideConfig,
     dashboardToken :: Text,
     cacheConfig :: CacheConfig,
+    minimumDriverRatesCount :: Int,
```
Comment (RitikaHanish) [inline Backend/app/rider-platform/rider-app/Main/src/Environment.hs:104]: Move this config to DB.

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Rating.hs [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+ratingToPSModifiers :: M.Map Text (A.Value -> A.Value)
+ratingToPSModifiers =
+  M.empty
+
+instance Serialize Rating where
+  put = error "undefined"
+  get = error "undefined"
+
+$(enableKVPG ''RatingT ['id] [['rideId]])
```
Comment (RitikaHanish) [inline Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Rating.hs:102]: riderId also should come in this.

Comment (RitikaHanish) [reply Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Rating.hs:102]: Can you check with Arjun Ashok once ?

File: Backend/dev/migrations/rider-app/1156-add-riderappbaseurl.sql [hunk 15]
```diff
@@ -0,0 +1 @@
+ALTER TABLE atlas_app.merchant ADD COLUMN internal_api_key character (255) DEFAULT 'test-api-key-for-dev-env' NOT NULL;
```
Comment (RitikaHanish) [inline Backend/dev/migrations/rider-app/1156-add-riderappbaseurl.sql:1]: it should be encrypted.

**Review by RitikaHanish — CHANGES_REQUESTED** [review]

File: Backend/dhall-configs/dev/secrets/dynamic-offer-driver-app.dhall [hunk 16]
```diff
@@ -9,4 +9,6 @@
 , clickHousePassword = "xxxx"
 , locDBUserId = "atlas_person_location_user"
 , locationTrackingServiceKey = "ae288466-2add-11ee-be56-0242ac120002"
+, appBackendApikey = "test-bap-api-key"
+, internalKey = "test-bpp-api-key"
```
Comment (RitikaHanish) [inline Backend/dhall-configs/dev/secrets/dynamic-offer-driver-app.dhall:13]: rename it to internalAPIKey

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Internal/Rating.hs [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+  calculateAverageRating booking.riderId
+  pure Success
+
+calculateAverageRating ::
+  (EsqDBFlow m r, EncFlow m r, HasFlowEnv m r '["minimumDriverRatesCount" ::: Int]) =>
+  Id DP.Person ->
+  m ()
+calculateAverageRating personId = do
+  logTagInfo "PersonAPI" $ "Recalculating average rating for driver " +|| personId ||+ ""
+  allRatings <- B.runInReplica $ QRating.findAllRatingsForPerson personId
```
Comment (RitikaHanish) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Internal/Rating.hs:84]: fetch person current RatingCount and Rating Score.
then Update ratingCount++, ratingScore = givenRating+currRatingScore

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Internal/Rating.hs [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+    CacheFlow m r,
+    EncFlow m r,
+    HasFlowEnv m r '["minimumDriverRatesCount" ::: Int],
+    HasFlowEnv m r '["driverOfferBppInternal" ::: DriverOfferBppInternal]
+  ) =>
+  Maybe Text ->
+  FeedbackReq ->
+  m APISuccess
+rating apiKey FeedbackReq {..} = do
+  driverOfferBppInternal <- asks (.driverOfferBppInternal)
```
Comment (RitikaHanish) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/Internal/Rating.hs:57]: Don't use this whole structure,  define internalAPIKey separately in Config and enviornment and use it.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs [hunk 19]
```diff
@@ -52,20 +56,20 @@ handler req ride = do
...
-  (EsqDBFlow m r, EncFlow m r, HasFlowEnv m r '["minimumDriverRatesCount" ::: Int]) =>
+  (EsqDBFlow m r, EncFlow m r) =>
   Id DP.Person ->
+  Int ->
   m ()
-calculateAverageRating personId = do
+calculateAverageRating personId minimumDriverRatesCount = do
   logTagInfo "PersonAPI" $ "Recalculating average rating for driver " +|| personId ||+ ""
-  allRatings <- QRating.findAllRatingsForPerson personId
+  allRatings <- B.runInReplica $ QRating.findAllRatingsForPerson personId
```
Comment (RitikaHanish) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Rating.hs:68]: Let's follow same approach as customer

File: Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Rating.hs [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+ratingToPSModifiers :: M.Map Text (A.Value -> A.Value)
+ratingToPSModifiers =
+  M.empty
+
+instance Serialize Rating where
+  put = error "undefined"
+  get = error "undefined"
+
+$(enableKVPG ''RatingT ['id] [['rideId], ['riderId]])
```
[the remark this answers is not in the document]
Comment (RitikaHanish) [reply Backend/app/rider-platform/rider-app/Main/src/Storage/Beam/Rating.hs:102]: @vsaimanohar Can you remove riderId from this mapping ?

**Review by RitikaHanish — APPROVED** [review]

**Review by hkmangla — APPROVED** [review]

**Comment by RitikaHanish** [comment]
@vsaimanohar Create another PR for customer rating driver refactoring as discussed.
